checkpoint module utils and service tcp module for example

chkp-orso opened the pull request:
##### SUMMARY
<!--- Describe the change below, including rationale and design decisions -->

<!--- HINT: Include "Fixes #nnn" if you are fixing an existing issue -->

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request
- Docs Pull Request
- Feature Pull Request
- New Module Pull Request

##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->

##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here -->
<!--- A step-by-step reproduction of the problem is helpful if there is no related issue -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```paste below

```

chkp-orso committed:
checkpoint module utols and service tcp module for example

chkp-orso committed:
Revert "Fortinet's FortiOS Authentication Rule (#51817)"

This reverts commit 6afa00eecd29c0937a2677f25a1aef8645539d37.

chkp-orso committed:
Revert "Revert "Fortinet's FortiOS Authentication Rule (#51817)""

This reverts commit 0bde0ff423d167119e8dd002db8c633d2508b309.

chkp-orso committed:
checkpoint_network module and checkpoint utils - example for module implementation

chkp-orso committed:
delete service tcp

chkp-orso committed:
Revert "delete service tcp"

This reverts commit e768efd146284d87504613b408f17688073449f0.

chkp-orso committed:
delete service tcp

chkp-orso committed:
delete network and module utils

chkp-orso committed:
checkpoint_network module and checkpoint utils - example for module implementation

chkp-orso committed:
Examples for modules network, network facts and inatsll policy. And the code of the module utils of checkpoint. Is this implementation is good enough from your aspect?

Examples for modules network, network facts and inatsll policy. And the code of the module utils of checkpoint. Is this implementation is good enough from your aspect?

justjais commented:
@chkp-orso can you please fix the shippable error before me reviewing the PR, also you can test these fixes locally before pushing the changes to the branch. Please follow the shippable errors [comment](https://github.com/ansible/ansible/pull/52932#issuecomment-480601066), and all the respective tests can be run locally.

justjais commented on lib/ansible/module_utils/network/checkpoint/checkpoint.py:
```diff
@@ -57,3 +60,125 @@ def install_policy(connection, policy_package, targets):
+                pass
+        elif code == 404:
+            code, response = create_api_call_object(connection, api_call_object, payload)
+
```
please handle the exception response, if the response code is not `200` then you can throw the error response like: `self.module.fail_json(msg=response))` for all the api_calls.

chkp-orso replied:
Hi, that's not an error, because I get to this code if the user want to add object (state='present'), and the object is not on the database of checkpoint, and that's why the code is 404, so we add the new object..

justjais replied:
@chkp-orso I understand that, but the thing is if there is exception response from function `create_api_call_object` you are not handling that response, and this is why `checkpoint_host` module was not throwing error response when u tested with `invalid ip_addr`, so this was the fix that I had added in [PR](https://github.com/ansible/ansible/pull/54890/files).

chkp-orso replied:
ohh I see now what you mean, but the response itself which I get from "code, response = create_api_call_object(connection, api_call_object, payload)"
is still not the response of checkpoint, just "'Object not found'"

justjais replied:
@chkp-orso I have verified the changes in the PR and the response exception in case of incorrect `ip_address`  was same as thrown from `checkpoint` api, also if u don't implement the respective check then in that case u'll receive `Object Not found` error from here: [AnsibleConnectionFailure](https://github.com/ansible/ansible/blob/97ac03f613bf747c07e526f012044323077162f1/lib/ansible/plugins/httpapi/checkpoint.py#L66)

chkp-orso replied:
@justjais so how can I get this response? I tried:

            code, response = add_api_call_object(connection, api_call_object, payload)
            if code is 404:
                module.fail_json(msg=response)

and I also tried:

            try:
                code, response = add_api_call_object(connection, api_call_object, payload)
            except Exception as e:
                module.fail_json(msg=response)

but I don't get the checkpoint response

justjais replied:
@chkp-orso u are not able to get the error response from CP, coz the response code is not `404` when incorrect IP address is given as input its `400`, change the if check as in the PR:
```
if code != 200:
  module.fail_json(msg=response)
```
and respective check shall handle all the error code other than success i.e. 200 and shall throw the respective CP error. Also the second code block won't work coz its not n exception its the response and the code that need to be checked for getting the exact response from CP.

chkp-orso replied:
@justjais ohh I see, I changed it but now I now I get only part of checkpoint output.
The output is a dictionary, and I get only the message key, I need somehow to get the whole dictionary output
Thanks!

justjais replied:
@chkp-orso, can you please install python `q` module and debug the `code n response` that u are getting once you are calling the CP api. The way you can debug the code is once u install the python `q` module u can import the `q` module in the code where you want to debug, then u can directly put debug statements via `q` module as:
`q("Debug Statement!")` and `q(code, response)`
and this should generate debug values under the TMPDIR plz refer following link for further info [q module](https://pypi.org/project/q/).
Using this u can detect what exactly is going wrong.

chkp-orso replied:
@justjais  Thanks I didn't know this module. 
So debug and this is the values of code and response:

      code=404,
      response=u'Could not connect to https://172.23.78.74:443/web_api/add-network: Bad Request'

and it's only the first part of the error request of checkpoint :(

justjais replied:
@chkp-orso can u run the same scenario which you are trying via some restclient and copy-paste the output here for me to understand the exact issue here.

chkp-orso replied:
@justjais If I run the same scenario (when subnet is illegal - 2.2.2.2) threw postman (web-api) the output is the following json:
{
    "code": "err_validation_failed",
    "message": "Validation failed with 1 warning",
    "warnings": [
        {
            "message": "The IPv4 address is invalid based on the defined net mask. The suggested address: 2.2.2.0 will be used instead."
        }
    ]
}

justjais replied:
@chkp-orso plz debug the `httpapi` connection plugin via `q` module as why there is difference in the response captured from `send_request`.

chkp-orso replied:
@justjais I tried it, but it's not recognize the q module.
The following error is appear: 

File "/home/admin/ansible/lib/ansible/plugins/httpapi/checkpoint.py", line 57, in send_request
    import q; q("send request")
ImportError: No module named q

justjais replied:
@chkp-orso it seems to be some python/Ansible environment issue, as I am not facing the respective issue and I am able to debug the respective lines. Can you please try to triage the respective issue from your end.

chkp-orso replied:
@justjais  I tried for a few days but didn't succeeded so far :(

chkp-orso replied:
maybe the cause of this problem is the rpc method, means `send_request` function is called from plugins/httpapi/__init__.py via the Connection object **rpc** method as per the ansible design

justjais commented on lib/ansible/module_utils/network/checkpoint/checkpoint.py:
```diff
@@ -57,3 +60,125 @@ def install_policy(connection, policy_package, targets):
     connection.send_request('/web_api/install-policy', payload)
+
+
+def get_api_call_object(connection, api_call_object, unique_payload_for_get):
```
please mention a `one-liner` comment for all the functions.

justjais commented on lib/ansible/modules/network/checkpoint/checkpoint_network.py:
```diff
@@ -0,0 +1,210 @@
+from ansible.module_utils.network.checkpoint.checkpoint import checkpoint_argument_spec, api_call
+
+
+argument_spec = dict(
```
please move the `argument_spec` under `main()`.

chkp-orso committed:
TRY 2: Examples for modules network, network facts and inatsll policy. And the code of the module utils of checkpoint. Is this implementation is good enough from your aspect?

TRY 2: Examples for modules network, network facts and inatsll policy. And the code of the module utils of checkpoint. Is this implementation is good enough from your aspect?

chkp-orso committed:
TRY 3: Examples for modules network, network facts and inatsll policy. And the code of the module utils of checkpoint. Is this implementation is good enough from your aspect?

TRY 3: Examples for modules network, network facts and inatsll policy. And the code of the module utils of checkpoint. Is this implementation is good enough from your aspect?

chkp-orso committed:
remove default from documentation

remove default from documentation
